README.md
# serverless-image-handler

Serverless image processing using AWS Lambda and [Sharp.js](https://sharp.pixelplumbing.com/).

## About this fork

This is a fork of [awslabs/serverless-image-handler](https://github.com/awslabs/serverless-image-handler).

The primary goal of this fork is to make the upstream project simpler to maintain and manage

Major changes:

* Removed the demo UI.
* Removed usage metrics reporting.
* Removed the additional `custom-resource` lambda.
* Hermetic build of source code zip file using Docker.

## Building

Build and test commands leverage Docker to ensure both run in an environment that as similar as possible to the AWS Lambda Node runtime.

To build:
```
make builder
```

To test:
```
make test
```

## Releasing

The CloudFormation template, and the code it refers to, must be released to an S3 bucket.

To deploy, you'll refer to the public URL of the template in this bucket.

Here's how to build and upload a release:

```
# Create and upload a release.
make release upload-release \
  DIST_S3_BUCKET_NAME=some-bucket-name \
  DIST_S3_BUCKET_REGION=us-east-1
```

The following values are passed to `make`:

* `DIST_S3_BUCKET_NAME`: An S3 bucket you maintain only for holding the source code and CloudFormation template. _It is not part of image processing_ and should be a different bucket.
* `DIST_S3_BUCKET_REGION`: The region this bucket was created in.

The command above will execute `aws s3api put-object` and whatever credentials are available for it in your environment.


CHANGELOG.md
# Changelog

## Current version (in development)

* Forked from [awslabs/serverless-image-handler@19cbc3ce](https://github.com/awslabs/serverless-image-handler/tree/19cbc3ce759d7c8d8ddc35081972d7ac1daf0c71).
* Removed the demo UI.
* Removed usage metrics reporting.
* Removed the additional `custom-resource` lambda.
* Use `yarn`.
* Use `eslint` and `prettier`.
* Add hermetic build using Docker (`make release`)
* Bugfix: Fixed handling of `AUTO_WEBP` env flag.


test/test-image-handler.js
/** *******************************************************************************************************************
 *  Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 ******************************************************************************************************************** */

const sharp = require("sharp");
const assert = require("assert");
const ImageHandler = require("../image-handler");

// ----------------------------------------------------------------------------
// [async] process()
// ----------------------------------------------------------------------------
describe("process()", () => {
  describe("001/default", () => {
    it("Should pass if the output image is different from the input image with edits applied", async () => {
      // Arrange
      const sinon = require("sinon");
      // ---- Amazon S3 stub
      const S3 = require("aws-sdk/clients/s3");
      const getObject = (S3.prototype.getObject = sinon.stub());
      getObject.returns({
        promise: () => ({
          Body: Buffer.from(
            "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg==",
            "base64"
          ),
        }),
      });
      // ----
      const request = {
        requestType: "default",
        bucket: "sample-bucket",
        key: "sample-image-001.jpg",
        edits: {
          grayscale: true,
          flip: true,
        },
        originalImage: Buffer.from(
          "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg==",
          "base64"
        ),
      };
      // Act
      const imageHandler = new ImageHandler();
      const result = await imageHandler.process(request);
      // Assert
      assert.deepEqual(request.originalImage !== result, true);
    });
  });
  describe("002/withToFormat", () => {
    it("Should pass if the output image is in a different format than the original image", async () => {
      // Arrange
      const sinon = require("sinon");
      // ---- Amazon S3 stub
      const S3 = require("aws-sdk/clients/s3");
      const getObject = (S3.prototype.getObject = sinon.stub());
      getObject.returns({
        promise: () => ({
          Body: Buffer.from(
            "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg==",
            "base64"
          ),
        }),
      });
      // ----
      const request = {
        requestType: "default",
        bucket: "sample-bucket",
        key: "sample-image-001.jpg",
        outputFormat: "png",
        edits: {
          grayscale: true,
          flip: true,
        },
        originalImage: Buffer.from(
          "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg==",
          "base64"
        ),
      };
      // Act
      const imageHandler = new ImageHandler();
      const result = await imageHandler.process(request);
      // Assert
      assert.deepEqual(request.originalImage !== result, true);
    });
  });
  describe("003/noEditsSpecified", () => {
    it("Should pass if no edits are specified and the original image is returned", async () => {
      // Arrange
      const sinon = require("sinon");
      // ---- Amazon S3 stub
      const S3 = require("aws-sdk/clients/s3");
      const getObject = (S3.prototype.getObject = sinon.stub());
      getObject.returns({
        promise: () => ({
          Body: Buffer.from(
            "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg==",
            "base64"
          ),
        }),
      });
      // ----
      const request = {
        requestType: "default",
        bucket: "sample-bucket",
        key: "sample-image-001.jpg",
        originalImage: Buffer.from(
          "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg==",
          "base64"
        ),
      };
      // Act
      const imageHandler = new ImageHandler();
      const result = await imageHandler.process(request);
      // Assert
      assert.deepEqual(
        result,
        "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg=="
      );
    });
  });
});

// ----------------------------------------------------------------------------
// [async] applyEdits()
// ----------------------------------------------------------------------------
describe("applyEdits()", () => {
  describe("001/standardEdits", () => {
    it(`Should pass if a series of standard edits are provided to the
            function`, async () => {
      // Arrange
      const originalImage = Buffer.from(
        "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg==",
        "base64"
      );
      const edits = {
        grayscale: true,
        flip: true,
      };
      // Act
      const imageHandler = new ImageHandler();
      const result = await imageHandler.applyEdits(originalImage, edits);
      // Assert
      const expectedResult1 = result.options.greyscale;
      const expectedResult2 = result.options.flip;
      const combinedResults = expectedResult1 && expectedResult2;
      assert.deepEqual(combinedResults, true);
    });
  });
  describe("002/overlay", () => {
    it(`Should pass if an edit with the overlayWith keyname is passed to
            the function`, async () => {
      // Arrange
      const sinon = require("sinon");
      // ---- Amazon S3 stub
      const S3 = require("aws-sdk/clients/s3");
      const getObject = (S3.prototype.getObject = sinon.stub());
      getObject.returns({
        promise: () => ({
          Body: Buffer.from(
            "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg==",
            "base64"
          ),
        }),
      });
      // Act
      const originalImage = Buffer.from(
        "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg==",
        "base64"
      );
      const edits = {
        overlayWith: {
          bucket: "aaa",
          key: "bbb",
        },
      };
      // Assert
      const imageHandler = new ImageHandler();
      await imageHandler.applyEdits(originalImage, edits).then((result) => {
        assert.deepEqual(result.options.input.buffer, originalImage);
      });
    });
  });
  describe("003/smartCrop", () => {
    it(`Should pass if an edit with the smartCrop keyname is passed to
            the function`, async () => {
      // Arrange
      const sinon = require("sinon");
      // ---- Amazon Rekognition stub
      const rekognition = require("aws-sdk/clients/rekognition");
      const detectFaces = (rekognition.prototype.detectFaces = sinon.stub());
      detectFaces.returns({
        promise: () => ({
          FaceDetails: [
            {
              BoundingBox: {
                Height: 0.18,
                Left: 0.55,
                Top: 0.33,
                Width: 0.23,
              },
            },
          ],
        }),
      });
      // Act
      const originalImage = Buffer.from(
        "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg==",
        "base64"
      );
      const edits = {
        smartCrop: {
          faceIndex: 0,
          padding: 0,
        },
      };
      // Assert
      const imageHandler = new ImageHandler();
      await imageHandler
        .applyEdits(originalImage, edits)
        .then((result) => {
          // console.log(result);
          const sharp = require("sharp");
          const originalImageData = sharp(originalImage);
          assert.deepEqual(
            originalImageData.options.input !== result.options.input,
            true
          );
        })
        .catch((err) => {
          console.log(err);
        });
    });
  });
  describe("004/smartCrop/paddingOutOfBoundsError", () => {
    it(`Should pass if an excessive padding value is passed to the
            smartCrop filter`, async () => {
      // Arrange
      const sinon = require("sinon");
      // ---- Amazon Rekognition stub
      const rekognition = require("aws-sdk/clients/rekognition");
      const detectFaces = (rekognition.prototype.detectFaces = sinon.stub());
      detectFaces.returns({
        promise: () => ({
          FaceDetails: [
            {
              BoundingBox: {
                Height: 0.18,
                Left: 0.55,
                Top: 0.33,
                Width: 0.23,
              },
            },
          ],
        }),
      });
      // Act
      const originalImage = Buffer.from(
        "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg==",
        "base64"
      );
      const edits = {
        smartCrop: {
          faceIndex: 0,
          padding: 80,
        },
      };
      // Assert
      const imageHandler = new ImageHandler();
      await imageHandler
        .applyEdits(originalImage, edits)
        .then((result) => {
          // console.log(result);
          const sharp = require("sharp");
          const originalImageData = sharp(originalImage);
          assert.deepEqual(
            originalImageData.options.input !== result.options.input,
            true
          );
        })
        .catch((err) => {
          console.log(err);
        });
    });
  });
  describe("005/smartCrop/boundingBoxError", () => {
    it(`Should pass if an excessive faceIndex value is passed to the
            smartCrop filter`, async () => {
      // Arrange
      const sinon = require("sinon");
      // ---- Amazon Rekognition stub
      const rekognition = require("aws-sdk/clients/rekognition");
      const detectFaces = (rekognition.prototype.detectFaces = sinon.stub());
      detectFaces.returns({
        promise: () => ({
          FaceDetails: [
            {
              BoundingBox: {
                Height: 0.18,
                Left: 0.55,
                Top: 0.33,
                Width: 0.23,
              },
            },
          ],
        }),
      });
      // Act
      const originalImage = Buffer.from(
        "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg==",
        "base64"
      );
      const edits = {
        smartCrop: {
          faceIndex: 10,
          padding: 0,
        },
      };
      // Assert
      const imageHandler = new ImageHandler();
      await imageHandler
        .applyEdits(originalImage, edits)
        .then((result) => {
          // console.log(result);
          const sharp = require("sharp");
          const originalImageData = sharp(originalImage);
          assert.deepEqual(
            originalImageData.options.input !== result.options.input,
            true
          );
        })
        .catch((err) => {
          console.log(err);
        });
    });
  });
  describe("006/smartCrop/faceIndexUndefined", () => {
    it(`Should pass if a faceIndex value of undefined is passed to the
            smartCrop filter`, async () => {
      // Arrange
      const sinon = require("sinon");
      // ---- Amazon Rekognition stub
      const rekognition = require("aws-sdk/clients/rekognition");
      const detectFaces = (rekognition.prototype.detectFaces = sinon.stub());
      detectFaces.returns({
        promise: () => ({
          FaceDetails: [
            {
              BoundingBox: {
                Height: 0.18,
                Left: 0.55,
                Top: 0.33,
                Width: 0.23,
              },
            },
          ],
        }),
      });
      // Act
      const originalImage = Buffer.from(
        "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg==",
        "base64"
      );
      const edits = {
        smartCrop: true,
      };
      // Assert
      const imageHandler = new ImageHandler();
      await imageHandler
        .applyEdits(originalImage, edits)
        .then((result) => {
          // console.log(result);
          const sharp = require("sharp");
          const originalImageData = sharp(originalImage);
          assert.deepEqual(
            originalImageData.options.input !== result.options.input,
            true
          );
        })
        .catch((err) => {
          console.log(err);
        });
    });
  });
});

// ----------------------------------------------------------------------------
// [async] getOverlayImage()
// ----------------------------------------------------------------------------
describe("getOverlayImage()", () => {
  describe("001/validParameters", () => {
    it(`Should pass if the proper bucket name and key are supplied,
            simulating an image file that can be retrieved`, async () => {
      // Arrange
      const S3 = require("aws-sdk/clients/s3");
      const sinon = require("sinon");
      const getObject = (S3.prototype.getObject = sinon.stub());
      getObject.withArgs({ Bucket: "validBucket", Key: "validKey" }).returns({
        promise: () => ({
          Body: Buffer.from(
            "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg==",
            "base64"
          ),
        }),
      });
      // Act
      const imageHandler = new ImageHandler();
      const metadata = await sharp(
        Buffer.from(
          "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg==",
          "base64"
        )
      ).metadata();
      const result = await imageHandler.getOverlayImage(
        "validBucket",
        "validKey",
        "100",
        "100",
        "20",
        metadata
      );
      // Assert
      assert.deepEqual(
        result,
        Buffer.from(
          "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAACXBIWXMAAAsSAAALEgHS3X78AAAADUlEQVQI12P4z8CQCgAEZgFlTg0nBwAAAABJRU5ErkJggg==",
          "base64"
        )
      );
    });
  });
  describe("002/imageDoesNotExist", async () => {
    it(`Should throw an error if an invalid bucket or key name is provided,
            simulating a non-existant overlay image`, async () => {
      // Arrange
      const S3 = require("aws-sdk/clients/s3");
      const sinon = require("sinon");
      const getObject = (S3.prototype.getObject = sinon.stub());
      getObject
        .withArgs({ Bucket: "invalidBucket", Key: "invalidKey" })
        .returns({
          promise: () =>
            Promise.reject({
              code: 500,
              message: "SimulatedInvalidParameterException",
            }),
        });
      // Act
      const imageHandler = new ImageHandler();
      // Assert
      imageHandler
        .getOverlayImage("invalidBucket", "invalidKey")
        .then((result) => {
          assert.equal(typeof result, Error);
        })
        .catch((err) => {
          console.log(err);
        });
    });
  });
});

// ----------------------------------------------------------------------------
// [async] getCropArea()
// ----------------------------------------------------------------------------
describe("getCropArea()", () => {
  describe("001/validParameters", () => {
    it(`Should pass if the crop area can be calculated using a series of
            valid inputs/parameters`, () => {
      // Arrange
      const boundingBox = {
        Height: 0.18,
        Left: 0.55,
        Top: 0.33,
        Width: 0.23,
      };
      const options = { padding: 20 };
      const metadata = {
        width: 200,
        height: 400,
      };
      // Act
      const imageHandler = new ImageHandler();
      const result = imageHandler.getCropArea(boundingBox, options, metadata);
      // Assert
      const expectedResult = {
        left: 90,
        top: 112,
        width: 86,
        height: 112,
      };
      assert.deepEqual(result, expectedResult);
    });
  });
});

// ----------------------------------------------------------------------------
// [async] getBoundingBox()
// ----------------------------------------------------------------------------
describe("getBoundingBox()", () => {
  describe("001/validParameters", () => {
    it("Should pass if the proper parameters are passed to the function", async () => {
      // Arrange
      const sinon = require("sinon");
      const rekognition = require("aws-sdk/clients/rekognition");
      const detectFaces = (rekognition.prototype.detectFaces = sinon.stub());
      // ----
      const imageBytes = Buffer.from("TestImageData");
      detectFaces.withArgs({ Image: { Bytes: imageBytes } }).returns({
        promise: () => ({
          FaceDetails: [
            {
              BoundingBox: {
                Height: 0.18,
                Left: 0.55,
                Top: 0.33,
                Width: 0.23,
              },
            },
          ],
        }),
      });
      // ----
      const currentImage = imageBytes;
      const faceIndex = 0;
      // Act
      const imageHandler = new ImageHandler();
      const result = await imageHandler.getBoundingBox(currentImage, faceIndex);
      // Assert
      const expectedResult = {
        Height: 0.18,
        Left: 0.55,
        Top: 0.33,
        Width: 0.23,
      };
      assert.deepEqual(result, expectedResult);
    });
  });
  describe("002/errorHandling", () => {
    it("Should simulate an error condition returned by Rekognition", async () => {
      // Arrange
      const rekognition = require("aws-sdk/clients/rekognition");
      const sinon = require("sinon");
      const detectFaces = (rekognition.prototype.detectFaces = sinon.stub());
      detectFaces.returns({
        promise: () =>
          Promise.reject({
            code: 500,
            message: "SimulatedError",
          }),
      });
      // ----
      const currentImage = Buffer.from("NotTestImageData");
      const faceIndex = 0;
      // Act
      const imageHandler = new ImageHandler();
      // Assert
      imageHandler
        .getBoundingBox(currentImage, faceIndex)
        .then((result) => {
          assert.equal(typeof result, Error);
        })
        .catch((err) => {
          console.log(err);
        });
    });
  });
});


image-request.js
/** *******************************************************************************************************************
 *  Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 ******************************************************************************************************************** */

const ThumborMapping = require("./thumbor-mapping");

class ImageRequest {
  /**
   * Initializer function for creating a new image request, used by the image
   * handler to perform image modifications.
   * @param {Object} event - Lambda request body.
   */
  async setup(event) {
    try {
      this.requestType = this.parseRequestType(event);
      this.bucket = this.parseImageBucket(event, this.requestType);
      this.key = this.parseImageKey(event, this.requestType);
      this.edits = this.parseImageEdits(event, this.requestType);
      this.originalImage = await this.getOriginalImage(this.bucket, this.key);

      /* Decide the output format of the image.
       * 1) If the format is provided, the output format is the provided format.
       * 2) If headers contain "Accept: image/webp", the output format is webp.
       * 3) Use the default image format for the rest of cases.
       */
      const outputFormat = this.getOutputFormat(event);
      if (this.edits && this.edits.toFormat) {
        this.outputFormat = this.edits.toFormat;
      } else if (outputFormat) {
        this.outputFormat = outputFormat;
      }

      // Fix quality for Thumbor and Custom request type if outputFormat is different from quality type.
      if (this.outputFormat) {
        const requestType = ["Custom", "Thumbor"];
        const acceptedValues = ["jpeg", "png", "webp", "tiff", "heif"];

        this.ContentType = `image/${this.outputFormat}`;
        if (
          requestType.includes(this.requestType) &&
          acceptedValues.includes(this.outputFormat)
        ) {
          const qualityKey = Object.keys(this.edits).filter((key) =>
            acceptedValues.includes(key)
          )[0];
          if (qualityKey && qualityKey !== this.outputFormat) {
            const qualityValue = this.edits[qualityKey];
            this.edits[this.outputFormat] = qualityValue;
            delete this.edits[qualityKey];
          }
        }
      }

      return Promise.resolve(this);
    } catch (err) {
      return Promise.reject(err);
    }
  }

  /**
   * Gets the original image from an Amazon S3 bucket.
   * @param {String} bucket - The name of the bucket containing the image.
   * @param {String} key - The key name corresponding to the image.
   * @return {Promise} - The original image or an error.
   */
  async getOriginalImage(bucket, key) {
    const S3 = require("aws-sdk/clients/s3");
    const s3 = new S3();
    const imageLocation = { Bucket: bucket, Key: key };
    try {
      const originalImage = await s3.getObject(imageLocation).promise();

      if (originalImage.ContentType) {
        this.ContentType = originalImage.ContentType;
      } else {
        this.ContentType = "image";
      }

      if (originalImage.Expires) {
        this.Expires = new Date(originalImage.Expires).toUTCString();
      }

      if (originalImage.LastModified) {
        this.LastModified = new Date(originalImage.LastModified).toUTCString();
      }

      if (originalImage.CacheControl) {
        this.CacheControl = originalImage.CacheControl;
      } else {
        this.CacheControl = "max-age=31536000,public";
      }

      return Promise.resolve(originalImage.Body);
    } catch (err) {
      return Promise.reject({
        status: err.code === "NoSuchKey" ? 404 : 500,
        code: err.code,
        message: err.message,
      });
    }
  }

  /**
   * Parses the name of the appropriate Amazon S3 bucket to source the
   * original image from.
   * @param {String} event - Lambda request body.
   * @param {String} requestType - Image handler request type.
   */
  parseImageBucket(event, requestType) {
    if (requestType === "Default") {
      // Decode the image request
      const decoded = this.decodeRequest(event);
      if (decoded.bucket !== undefined) {
        // Check the provided bucket against the whitelist
        const sourceBuckets = this.getAllowedSourceBuckets();
        if (
          sourceBuckets.includes(decoded.bucket) ||
          decoded.bucket.match(new RegExp(`^${sourceBuckets[0]}$`))
        ) {
          return decoded.bucket;
        }
        throw {
          status: 403,
          code: "ImageBucket::CannotAccessBucket",
          message:
            "The bucket you specified could not be accessed. Please check that the bucket is specified in your SOURCE_BUCKETS.",
        };
      } else {
        // Try to use the default image source bucket env var
        const sourceBuckets = this.getAllowedSourceBuckets();
        return sourceBuckets[0];
      }
    } else if (requestType === "Thumbor" || requestType === "Custom") {
      // Use the default image source bucket env var
      const sourceBuckets = this.getAllowedSourceBuckets();
      return sourceBuckets[0];
    } else {
      throw {
        status: 404,
        code: "ImageBucket::CannotFindBucket",
        message:
          "The bucket you specified could not be found. Please check the spelling of the bucket name in your request.",
      };
    }
  }

  /**
   * Parses the edits to be made to the original image.
   * @param {String} event - Lambda request body.
   * @param {String} requestType - Image handler request type.
   */
  parseImageEdits(event, requestType) {
    if (requestType === "Default") {
      const decoded = this.decodeRequest(event);
      return decoded.edits;
    }
    if (requestType === "Thumbor") {
      const thumborMapping = new ThumborMapping();
      thumborMapping.process(event);
      return thumborMapping.edits;
    }
    if (requestType === "Custom") {
      const thumborMapping = new ThumborMapping();
      const parsedPath = thumborMapping.parseCustomPath(event.path);
      thumborMapping.process(parsedPath);
      return thumborMapping.edits;
    }
    throw {
      status: 400,
      code: "ImageEdits::CannotParseEdits",
      message:
        "The edits you provided could not be parsed. Please check the syntax of your request and refer to the documentation for additional guidance.",
    };
  }

  /**
   * Parses the name of the appropriate Amazon S3 key corresponding to the
   * original image.
   * @param {String} event - Lambda request body.
   * @param {String} requestType - Type, either "Default", "Thumbor", or "Custom".
   */
  parseImageKey(event, requestType) {
    if (requestType === "Default") {
      // Decode the image request and return the image key
      const decoded = this.decodeRequest(event);
      return decoded.key;
    }

    if (requestType === "Thumbor" || requestType === "Custom") {
      return decodeURIComponent(
        event.path
          .replace(/\d+x\d+\/|filters[:-][^/;]+|\/fit-in\/+|^\/+/g, "")
          .replace(/^\/+/, "")
      );
    }

    // Return an error for all other conditions
    throw {
      status: 404,
      code: "ImageEdits::CannotFindImage",
      message:
        "The image you specified could not be found. Please check your request syntax as well as the bucket you specified to ensure it exists.",
    };
  }

  /**
   * Determines how to handle the request being made based on the URL path
   * prefix to the image request. Categorizes a request as either "image"
   * (uses the Sharp library), "thumbor" (uses Thumbor mapping), or "custom"
   * (uses the rewrite function).
   * @param {Object} event - Lambda request body.
   */
  parseRequestType(event) {
    const { path } = event;
    // ----
    const matchDefault = new RegExp(
      /^(\/?)([0-9a-zA-Z+\/]{4})*(([0-9a-zA-Z+\/]{2}==)|([0-9a-zA-Z+\/]{3}=))?$/
    );
    const matchThumbor = new RegExp(
      /^(\/?)((fit-in)?|(filters:.+\(.?\))?|(unsafe)?).*(.+jpg|.+png|.+webp|.+tiff|.+jpeg)$/i
    );
    const matchCustom = new RegExp(/(\/?)(.*)(jpg|png|webp|tiff|jpeg)/i);
    const definedEnvironmentVariables =
      process.env.REWRITE_MATCH_PATTERN !== "" &&
      process.env.REWRITE_SUBSTITUTION !== "" &&
      process.env.REWRITE_MATCH_PATTERN !== undefined &&
      process.env.REWRITE_SUBSTITUTION !== undefined;
    // ----
    if (matchDefault.test(path)) {
      // use sharp
      return "Default";
    }
    if (matchCustom.test(path) && definedEnvironmentVariables) {
      // use rewrite function then thumbor mappings
      return "Custom";
    }
    if (matchThumbor.test(path)) {
      // use thumbor mappings
      return "Thumbor";
    }
    throw {
      status: 400,
      code: "RequestTypeError",
      message:
        "The type of request you are making could not be processed. Please ensure that your original image is of a supported file type (jpg, png, tiff, webp) and that your image request is provided in the correct syntax. Refer to the documentation for additional guidance on forming image requests.",
    };
  }

  /**
   * Decodes the base64-encoded image request path associated with default
   * image requests. Provides error handling for invalid or undefined path values.
   * @param {Object} event - The proxied request object.
   */
  decodeRequest(event) {
    const { path } = event;
    if (path !== undefined) {
      const splitPath = path.split("/");
      const encoded = splitPath[splitPath.length - 1];
      const toBuffer = Buffer.from(encoded, "base64");
      try {
        // To support European characters, 'ascii' was removed.
        return JSON.parse(toBuffer.toString());
      } catch (e) {
        throw {
          status: 400,
          code: "DecodeRequest::CannotDecodeRequest",
          message:
            "The image request you provided could not be decoded. Please check that your request is base64 encoded properly and refer to the documentation for additional guidance.",
        };
      }
    } else {
      throw {
        status: 400,
        code: "DecodeRequest::CannotReadPath",
        message:
          "The URL path you provided could not be read. Please ensure that it is properly formed according to the solution documentation.",
      };
    }
  }

  /**
   * Returns a formatted image source bucket whitelist as specified in the
   * SOURCE_BUCKETS environment variable of the image handler Lambda
   * function. Provides error handling for missing/invalid values.
   */
  getAllowedSourceBuckets() {
    const sourceBuckets = process.env.SOURCE_BUCKETS;
    if (sourceBuckets === undefined) {
      throw {
        status: 400,
        code: "GetAllowedSourceBuckets::NoSourceBuckets",
        message:
          "The SOURCE_BUCKETS variable could not be read. Please check that it is not empty and contains at least one source bucket, or multiple buckets separated by commas. Spaces can be provided between commas and bucket names, these will be automatically parsed out when decoding.",
      };
    } else {
      const formatted = sourceBuckets.replace(/\s+/g, "");
      const buckets = formatted.split(",");
      return buckets;
    }
  }

  /**
   * Return the output format depending on the accepts headers and request type
   * @param {Object} event - The request body.
   */
  getOutputFormat(event) {
    const autoWebP = process.env.AUTO_WEBP == "Yes";
    if (
      autoWebP &&
      event.headers.Accept &&
      event.headers.Accept.includes("image/webp")
    ) {
      return "webp";
    }
    if (this.requestType === "Default") {
      const decoded = this.decodeRequest(event);
      return decoded.outputFormat;
    }

    return null;
  }
}

// Exports
module.exports = ImageRequest;
